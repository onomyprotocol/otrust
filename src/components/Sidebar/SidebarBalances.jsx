import React from 'react';
import styled from 'styled-components';

import { PrimaryButton } from 'components/Modals/styles';
import { responsive } from 'theme/constants';
import { CloseIcon } from './SidebarIcons';

const Balances = styled.div`
  display: flex;
  flex-direction: column;
  gap: 22px;

  padding: 40px;

  border-bottom: 1px solid ${props => props.theme.colors.bgHighlightBorder};

  @media screen and (max-width: ${responsive.laptop}) {
    padding: 24px;
  }

  @media screen and (max-width: ${responsive.tablet}) {
    flex-direction: row;
    align-items: center;
    grid-column: 1/3;
    grid-row: 2/3;

    border: none;
  }

  @media screen and (max-width: ${responsive.tabletSmall}) {
    gap: 40px;
  }

  @media screen and (max-width: ${responsive.smartphoneLarge}) {
    padding: 20px;
  }

  strong {
    display: block;
    margin-bottom: 10px;

    font-weight: 400;
    color: ${props => props.theme.colors.textSecondary};
  }
`;

const BalanceNumber = styled.div`
  font-family: 'Bebas Neue', sans-serif;
  font-size: 28px;
  font-weight: ${props => (props.strong ? 700 : 400)};

  > small {
    margin-left: 5px;

    font-family: 'Poppins', sans-serif;
    color: ${props => props.theme.colors.textSecondary};
    font-size: 11px;
  }

  @media screen and (max-width: ${responsive.laptop}) {
    font-size: 20px;
  }
`;

const Balance = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
`;

const BalancePrice = styled.div``;
const BalanceHint = styled.div``;

const SecondaryIcon = styled.a`
  display: flex;
  align-items: center;
  justify-content: center;

  height: 40px;
  width: 40px;

  background-color: ${props => props.theme.colors.bgHighlightBorder};
  border-radius: 8px;

  font-size: 20px;
  font-weight: 500;
  color: ${props => props.theme.colors.iconsSecondary};

  cursor: pointer;

  @media screen and (max-width: ${responsive.laptop}) {
    width: 32px;
    height: 32px;

    font-size: 14px;
  }

  @media screen and (max-width: ${responsive.tablet}) {
    margin-left: 24px;
  }

  &:hover {
    background-color: ${props => props.theme.colors.bgHighlightBorder_lighten};
  }

  &:active {
    background-color: ${props => props.theme.colors.bgHighlightBorder_darken};
  }
`;

const Approved = styled.div`
  display: inline-block;

  padding: 4px 30px 4px 8px;
  margin-top: 6px;

  position: relative;

  background-color: #2a303f;
  border-radius: 6px;

  color: ${props => props.theme.colors.highlightBlue};
  font-size: 10px;
  font-weight: 400;
  font-family: Poppins, sans-serif;

  @media screen and (max-width: ${responsive.tablet}) {
    margin: 0 0 0 12px;
  }

  svg {
    position: absolute;
    right: 4px;
    top: 2px;

    cursor: pointer;
    * {
      fill: currentColor;
    }
  }
`;

const WithdrawBtnWrapper = styled.div`
  margin-top: 20px;
  min-width: 150px;

  @media screen and (max-width: ${responsive.tablet}) {
    margin: 0 0 0 auto;
  }
`;

export default function SidebarBalances({ strong, weak, strongBalance, weakBalance }) {
  return (
    <Balances>
      <Balance>
        <BalancePrice>
          <strong>{strong} Balance</strong>
          <BalanceNumber>
            {strongBalance}
            <small> = $16,208.04</small>
          </BalanceNumber>
        </BalancePrice>
        <BalanceHint>
          <SecondaryIcon>i</SecondaryIcon>
        </BalanceHint>
      </Balance>
      <Balance>
        <BalancePrice>
          <strong>{weak} Balance</strong>
          <BalanceNumber strong>
            {weakBalance}
            <small> = $16,208.04</small>
            <Approved>
              <span>1,042 approved</span>
              <CloseIcon onClick={() => {}} />
            </Approved>
          </BalanceNumber>
        </BalancePrice>
        <BalanceHint>
          <SecondaryIcon>i</SecondaryIcon>
        </BalanceHint>
      </Balance>

      <WithdrawBtnWrapper>
        <PrimaryButton style={{ width: '100%' }}>Withdraw wNOM</PrimaryButton>
      </WithdrawBtnWrapper>
    </Balances>
  );
}
