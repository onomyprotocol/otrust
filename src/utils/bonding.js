const a = 100000000;

export function priceAtSupply(supply) {
  return (supply / a) ** 2;
}

// Bonding curve price at specified supply
export function supplyAtPrice(price) {
  //#NOM Sold = sqrt(ETH/NOM) * a
  return Math.sqrt(price) * a;
}

// NOM supply range to ETH
export function NOMsupplyETH(supplyTop, supplyBot) {
  // Integrate over curve to get amount of ETH needed to buy amount of NOM
  // ETH = a/3((supplyNOM_Top/a)^3 - (supplyNOM_Bot/a)^3)
  return (a / 3) * ((supplyTop / a) ** 3 - (supplyBot / a) ** 3);
}

// Returns Buy Quote for the purchase of NOM based on amount of ETH
// Parameters:
// Input
// amount: amount of ETH
// Output
// output: amount of NOM
export function ETHtoNOM(amount, supply) {
  const price = priceAtSupply(supply);
  // 1. Determine supply bottom
  // Price bottom is 1% above priceBondCurve
  const priceBot = price * 1.01;
  const supplyBot = supplyAtPrice(priceBot);
  // 2. Integrate over curve, and solve for supply top
  // (3*ETH/a + (supplyNOM_Bot/a)^3)^(1/3)
  const supplyTop = a * ((3 * amount) / a + supplyBot ** 3 / a ** 3) ** (1 / 3);
  // 3. Subtract supply bottom from top to get amount NOM for amount ETH
  const diff = supplyTop - supplyBot;
  return { supplyBot, supplyTop, diff };
}

// Returns Sell Quote: NOM for ETH (Dec 18)
// Parameters:
// Input
// uint256 amountNOM: amount of NOM to be sold (18 decimal)
// Output
// uint256: amount of ETH paid in Wei or ETH (18 decimal)
export function NOMtoETH(amount, supplyTop) {
  const supplyBot = supplyTop - amount;
  const diff = 0.99 * NOMsupplyETH(supplyTop, supplyBot);
  return { supplyBot, supplyTop, diff };
}
