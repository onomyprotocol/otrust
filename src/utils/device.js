const size = {
  mobileS: '320px',
  mobileM: '375px',
  mobileL: '425px',
  tablet: '768px',
  laptop: '1024px',
  desktopS: '1366px',
  desktopM: '1920px',
  desktopL: '2560px',
};

export const device = {
  mobileS: `(min-width: ${size.mobileS})`,
  mobileM: `(min-width: ${size.mobileM})`,
  mobileL: `(min-width: ${size.mobileL})`,
  tablet: `(min-width: ${size.tablet})`,
  laptop: `(min-width: ${size.laptop})`,
  desktopS: `(min-width: ${size.desktopS})`,
  desktopM: `(min-width: ${size.desktopM})`,
  desktopL: `(min-width: ${size.desktopL})`,
};
