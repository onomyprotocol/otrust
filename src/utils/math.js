import { BigNumber } from 'bignumber.js';

export function chopFloat(value, precision) {
  return Math.round(value * 10 ** precision) / 10 ** precision;
}

export function randomInt(min, max) {
  return Math.floor(Math.random() * (max - min + 1) + min);
}

export function truncate(str, maxDecimalDigits) {
  if (str.includes('.')) {
    const parts = str.split('.');
    return parts[0] + '.' + parts[1].slice(0, maxDecimalDigits);
  }
  return str;
}

export function format18(bignumber) {
  return bignumber.div(new BigNumber(10 ** 18));
}

export function parse18(bignumber) {
  return bignumber.times(new BigNumber(10 ** 18));
}

export var isNumber = function isNumber(value) {
  return typeof value === 'number' && isFinite(value);
};
