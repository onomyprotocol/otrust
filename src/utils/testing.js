import { ThemeProvider } from 'styled-components';
import ApolloClient, { InMemoryCache } from 'apollo-boost';
import { ApolloProvider } from '@apollo/client';
import { render } from '@testing-library/react';
import { BigNumber } from 'bignumber.js';

import { darkNew } from 'Theme/theme';
import { ChainContext } from '../context/chain/ChainContext';
import { ExchangeContext } from '../context/exchange/ExchangeContext';

const client = new ApolloClient({
  uri: process.env.REACT_APP_GRAPHQL_ENDPOINT,
  cache: new InMemoryCache(),
});

export const renderWithTheme = (Component, props, children) => {
  if (children) {
    return render(
      <ThemeProvider theme={darkNew}>
        <Component {...props}>{children}</Component>
      </ThemeProvider>,
    );
  }
  return render(
    <ThemeProvider theme={darkNew}>
      <Component {...props} />
    </ThemeProvider>,
  );
};

export const renderWithContext = (Component, props) => {
  return render(
    <ThemeProvider theme={darkNew}>
      <ApolloProvider client={client}>
        <ChainContext.Provider value={{ supplyNOM: new BigNumber(0), theme: darkNew }}>
          <ExchangeContext.Provider value={{ askAmount: 0, bidAmount: 0, bidDenom: 'strong' }}>
            <Component {...props} />
          </ExchangeContext.Provider>
        </ChainContext.Provider>
      </ApolloProvider>
    </ThemeProvider>,
  );
};
