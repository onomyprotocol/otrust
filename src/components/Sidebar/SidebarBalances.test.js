import { cleanup } from '@testing-library/react';

import SidebarBalances from './SidebarBalances';
import { renderWithTheme } from '../../utils/testing';

const testProps = {
  strong: 'ETH',
  weak: 'NOM',
  strongBalance: 100,
  weakBalance: 50,
};

describe('Given the SidebarBalances component', () => {
  describe('when the component is rendered', () => {
    afterEach(cleanup);

    it('should match the snapshot', () => {
      const { asFragment } = renderWithTheme(SidebarBalances, testProps);
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
