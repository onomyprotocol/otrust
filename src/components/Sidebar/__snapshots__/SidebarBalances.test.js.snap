// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the SidebarBalances component when the component is rendered should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-kLojOw iGHpAL"
  >
    <div
      class="sc-jJMGnK cwMWz"
    >
      <div
        class="sc-hiKfDv"
      >
        <strong>
          ETH Balance
        </strong>
        <div
          class="sc-iklJeh bswktC"
        >
          100
          <small>
             = $16,208.04
          </small>
        </div>
      </div>
      <div
        class="sc-gXfVKN"
      >
        <a
          class="sc-cBoqAE hsnkhy"
        >
          i
        </a>
      </div>
    </div>
    <div
      class="sc-jJMGnK cwMWz"
    >
      <div
        class="sc-hiKfDv"
      >
        <strong>
          NOM Balance
        </strong>
        <div
          class="sc-iklJeh kmvZaR"
        >
          50
          <small>
             = $16,208.04
          </small>
          <div
            class="sc-ciSkZP dqkmvq"
          >
            <span>
              1,042 approved
            </span>
            <svg
              height="20"
              viewBox="0 0 20 20"
              width="20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g
                transform="translate(5 5.833)"
              >
                <path
                  class="cls-2"
                  d="M7.3.32a1.093 1.093 0 011.652 1.423l-.106.123-6.981 6.981A1.093 1.093 0 01.214 7.424L.32 7.3z"
                />
                <path
                  class="cls-2"
                  d="M.32.32A1.092 1.092 0 011.742.214l.122.106L8.847 7.3a1.092 1.092 0 01-1.422 1.65L7.3 8.847.32 1.864A1.092 1.092 0 01.32.32z"
                />
              </g>
            </svg>
          </div>
        </div>
      </div>
      <div
        class="sc-gXfVKN"
      >
        <a
          class="sc-cBoqAE hsnkhy"
        >
          i
        </a>
      </div>
    </div>
    <div
      class="sc-jcwpoC jwrGsS"
    >
      <button
        class="sc-hKFxyN hQFwnH"
        style="width: 100%;"
      >
        Withdraw wNOM
      </button>
    </div>
  </div>
</DocumentFragment>
`;
